mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -193,6 +200,44 @@ uploadSingle file a = do
   x <- liftIO $ LBS.readFile file
   void . send $ fencode' putObject a (toBody x) & poServerSideEncryption .~ pure sse
 
+-- | Upload a file from a URL to a given S3 Address
+uploadFromURL :: URL -> Address -> AWS UploadResult
+uploadFromURL = uploadFromURLWithMode Fail
+
+-- | Upload a file from a URL to a given S3 Address
+--   fail if the destination already exists and the WriteMode is Fail
+uploadFromURLWithMode :: WriteMode -> URL -> Address -> AWS UploadResult
+uploadFromURLWithMode m url a = eitherT (pure . UploadError) (const $ pure UploadOk) $ do
+  when (m == Fail) . whenM (lift $ exists a) . left $ UploadDestinationExists a
+  EitherT (Right <$> do
+    manager  <- liftIO $ newManager tlsManagerSettings
+    request  <- liftIO $ parseUrl (exportURL url)
+    response <- http request manager
+    let body =  responseBody response
+    let chunk = 100 * 1024 * 1024
+    body $$+- uploadSink a chunk)
+
+-- | Create an upload sink for a given S3 Address
+uploadSink :: Address -> Int -> Sink BS.ByteString AWS ()
+uploadSink a chunk = chunksOfAtLeast chunk =$ do
+  e <- lift ask
+  hoist liftIO $ do
+    bs <- C.await
+    maybe (pure ()) (liftIO . sendRequest e) bs
+    where
+      sendRequest :: Env -> LBS.ByteString -> IO ()
+      sendRequest e bs = runAWS e $ void . send $ fencode' putObject a (toBody bs) & poServerSideEncryption .~ pure sse
```

Comment:
Oh, and which runAWS is that, need to make sure:
 - It definitely isn't calling discover
 - It isn't recreating a http manager
 - it isn't losing retry policy

(I think this one is safe to all those things - because of propagating the e, just saying it out loud)